src/globus_sdk/services/auth/id_token_decoder.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import annotations
+
+import abc
+import datetime
+import typing as t
+
+import jwt
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPublicKey
+
+from globus_sdk.response import GlobusHTTPResponse
+
+from ._common import SupportsJWKMethods
+
+if t.TYPE_CHECKING:
+    from globus_sdk import AuthLoginClient, GlobusAppConfig
+
+
+class JWTDecoder(abc.ABC):
+    """
+    This abstract class defines a decoder for parsing JWTs.
+
+    A decoder is an object with a ``decode()`` method which accepts a token as its only
+    required (positional) parameter and produces a dict of decoded data.
+    """
+
+    @abc.abstractmethod
+    def decode(self, token: str, /) -> dict[str, t.Any]: ...
+
+
+class IDTokenDecoder(JWTDecoder):
+    """
+    An implementation of JWT decoding specialized to OIDC ID tokens issued by Globus
+    Auth. Decoding uses a client object to fetch necessary data from Globus Auth.
+
+    By default, the OIDC configuration data and JWKs will be cached in an internal dict.
+    An alternative cache can be provided on init to use an alternative storage
+    mechanism.
+
+    The ``get_jwt_audience`` and ``get_jwt_leeway`` methods supply parameters to
+    decoding. Subclasses can override these methods to customize the decoder.
+
+    :param auth_client: The client which should be used to callout to Globus Auth as
+        needed. Any AuthClient or AuthLoginClient will work for this purpose.
+
+    :ivar float | datetime.timedelta jwt_leeway: The JWT leeway to use during decoding,
+        as a number of seconds or a timedelta. The default is 5 minutes.
+    :ivar dict jwt_options: The ``options`` passed to the underlying JWT decode
+        function. Defaults to an empty dict.
+    """
+
+    def __init__(self, auth_client: SupportsJWKMethods) -> None:
```

Comment:
```suggestion
    def __init__(
        self,
        auth_client: SupportsJWKMethods,
        *,
        jwt_leeway: float = 300.0,  # seconds
        jwt_options: dict[str, t.Any] | MissingType = Missing,
    ) -> None:
```

Reply:
How would you feel about making these instance vars configurable as init params?

This makes jwt leeway default a part of the class signature, not just an internal comment in the code & allows for some more interesting extension patterns if users want to change it.

Extension 1:
```python
decoder = IDTokenDecoder(auth_client, jwt_leeway=10.0)
config = GlobusAppConfig(id_token_decoder=decoder)
```

Extension 2:
```python
config = GlobusAppConfig(
  id_token_decoder=functools.partial(IDTokenDecoder, jwt_leeway=10.0)
)
```

Reply:
Yeah, I like that. Let me refine to allow for it.

It might also simplify one specific odd mock for a test.